### ISSUE TITLE
Proposal: Marketing emphasis on RVM

### ISSUE THREAD

**Issue by moschel** [canjs/canjs#4053]
## Motivation

I'm thinking we have a good opportunity to add value to the React tools ecosystem using RVM as a React framework. It could compete against mobx and redux, and offer a competing paradigm. 

State management sucks (is very primitive) by default in React, and since the early days of flux it seems not much has improved, despite new tools. People writing React apps are still passing events around manually from component to component through dispatcher layers, and writing a ton of code where computes and bindings make little or no code necessary.

We already have a great start with React View Model, and it might just need some finishing touches and a marketing push to gain some adoption and drive people to using the CanJS ecosystem tools.

This ticket is to gauge interest in this idea and discuss what tangible steps would be needed to move this forward.

A few reasons for this:
 * The react community is just so huge, and yet still seems so fragmented and lacking - https://stateofjs.com/2017/front-end/results and https://stateofjs.com/2017/state-management/results/
 * Christopher recently gave a presentation about RVM, which left me with the impression that RVM is actually pretty much ready to go without much more work
 * It's being used by at least one of Bitovi's big clients

## Next steps

I think it might be worth it to put more effort into branding and promoting it:
 * Standalone homepage
 * Content around why it's better to have powerful viewmodels when building a react application and comparing that to flux/redux data flow - articles and videos
 * Possibly a better name - rvm sounds like ruby package manager
 * Docs can link back to the canjs pages for now

## Feedback

Please vote on this if you would like to prioritize this effort, and provide feedback on the following:
 * Is RVM good enough to start promoting as is? Or does it need some technical or design changes - and if so, what?
 * Should we re-brand it and promote it as a standalone framework with it's own landing page?

cc @BigAB @christopherjbaker @mickmcgrath13 @justinbmeyer

**Comment by BigAB** [comment]
I think the primary "syntactical direction" is somewhat still in flux at this moment (today at least).

1. I want to push the `connect()` HoC way of doing things, but there is some concerns about performance implications.

2. There is some discussion about changing the way react-components are connected to the view-models, primarily involving discussion around separating the props received from the component from the view model (a few different ways are being discussed)

3. I have an outlandish idea that is pretty much disliked by everyone, but I think may be the future:
```
const VMComponent = createComponentFromViewModel( ViewModel );

class UserProfile extends Component {
  render() {
    return <VMComponent render={ vm => {
        <figure>
          <img src={ vm.avatarURL } />
          <figcaption>
             <span>{ vm.fullname }</span>
             <span>{ vm.score} </span>
          </figcaption>
        </figure>
    } />
  }
}
```

**Long story short**, I think that we should start pushing it, and give it a better name, and a homepage and some videos, but we need to decide which way we are going to be pushing people to use it: extend `Component`, the `connect()` HoC/Decorator, a `VMComponent` with a render prop, or just a component created from a function that takes a `ViewModel` class and a special `render` function... and any other ideas that are currently on the table.

But **I don't think we have to wait for this decision** either, we can pick a new name, create a website and start planning some stuff out about how we will promote it while we decide on a syntax for creating connected components.  ...but in the end a decision has to be made before we can really start pushing it.

**Comment by christopherjbaker** [comment]
Though I think much of this was said by @BigAB, I think it is worth repeating in my own words. TLDR at bottom.

I think we should absolutely do the first 3 of your bullet points under "next steps" (see below for notes on the 4th bullet point). There are some changes that still need to be made, but I don't think any of these prevent moving forward, they just need to be settled before we "release" it.

## Docs

I think the RVM page should have basic docs self-located. Function signatures, example uses, etc. Those can link to equivalent pages in the canjs docs, which tend to be much more technical, which is perfect. I think directly them straight there would lose some (if for no other reason than they can get lost in the sea of canjs).

I think there should be 1 doc page for `can-observe` with basic usage + decorators, linking back to canjs for more details/examples/building decorators. `can-observe` is so vital to rvm that new users will be required to read a bit about it to do anything beyond the demos, and my same argument above applies.

In addition to "we are better than X because" (with whatever level of tact is decided), there should also be "switch to RVM from X like this" (perhaps "upgrate to RVM from X" instead?).

## Signatures

* The current stable RVM uses extending syntax. I think this method should be abandoned, though it can certainly be preserved in a deprecated state for a while (this is what the current client is using).
* There is an `@connect` branch that provides a Higher-Order Component interface, generally through a decorator (though a function call works just as well). I think this should be the default as presented in docs/etc.
* Adam has proposed a provider (`VMComponent`) interface that is very similar to the new way that react handles context (example can be found [here](https://medium.com/dailyjs/reacts-%EF%B8%8F-new-context-api-70c9fe01596b), scrolling down to the first gray code box). I think this is a good idea, though I think the structure should be reconciled with react context (as this is an established react pattern), but I don't think it would change much. This method provides distinct features over the `@connect` syntax, making it very clear when to use one over the other. This would be the advanced form.

My reasoning for rejecting `extend` in favor or HOC or providers is to make the react component itself as plain-react as possible, with no special knowledge. That way the VM can be connected to to any existing component (even 3rd party components) with 0 changes, and people using other state management systems like redux won't have to change their components at all, only create VMs that mimic their previous state files and change their `xxHandler` calls to vm calls (IE: instead of `<div onClick={() => this.clickHandler()) />`, where `clickHandler` dispatches an action, you do `<div onClick={this.props.clickHandler} />, where `clickHandler` on the vm just does' whatever it needs to do).

## Ch-ch-ch-ch-Changes (Turn and face the strange...)

For the `@connect` syntax, there is a performance problem that needs to be resolved. Basically, the viewmodel instance is shallow-cloned (and all methods are bound) when it is passed into the react app. I am quite sure I can resolve this without much effort. While this needs to be fixed, it will not change the usage at all. The docs for this need to be finished.

@justinbmeyer proposed passing the props to the viewmodel in such a way that the input props and any derived props and state were separate (react has a distinction between `this.props` and `this.state`). While I see the logic, I do not think this is a good idea. I do think that it is a good idea to point out how this (and other "don't merge props with vm" styles) can be accomplished (last point below) in an "advanced techniques" section.
1. If you were to use those values without derived getters, your jsx might look like `<div>{this.props.props.name}</div>`
1. I do not think the original source of information is relevant, only that the information exists for the component.
1. None of RVMs competitors, AFAIK, do this and I would like to make adoption as easy as possible (especially from redux, as it is the most popular option for state management besides "nothing", which has almost twice as many users as redux).
1. It is rather easy to do this yourself in the can-observe constructors if you prefer it. I found in my own apps that I often override the default "merge all props into the VM" with custom locations anyway.

## TLDR

1. New name
1. Standalone site with basic docs on RVM and a single page of `can-observe`.
1. Use `@connect` as default. have provider interface (mimicking reacts new context interface) for advanced usage. Abandon `extend`.
1. Fix performance issue in `@connect` and finish docs.
1. Build out provider/consumer components with docs.
1. More docs, including advanced usage and comparison/upgrade vs competitors (redux, and maybe mobx).

**Comment by justinbmeyer** [comment]
@christopherjbaker A few thoughts and questions on these points:

> 1. If you were to use those values without derived getters, your jsx might look like <div>{this.props.props.name}</div>

How could this happen?  I think you are talking about if a component passed itself.  My guess is you would probably see `<div>{this.props.passedComponent.props.name}</div>`.

> 2. I do not think the original source of information is relevant, only that the information exists for the component.

Both react (having `.props` and `.state`) and vuejs think this is important.   VueJS:

```js
Vue.component('item', {
  template: '#item-template',
  props: {
    model: Object
  },
  data: function () {
    return {
      open: false
    }
  },
  computed: {
    isFolder: function () {
      return this.model.children &&
        this.model.children.length
    }
  },
  methods: {
    toggle: function () {
      if (this.isFolder) {
        this.open = !this.open
      }
    },
    changeType: function () {
      if (!this.isFolder) {
        Vue.set(this.model, 'children', [])
        this.addChild()
        this.open = true
      }
    },
    addChild: function () {
      this.model.children.push({
        name: 'new stuff'
      })
    }
  }
})
```

It might not be relevant to the view.  But it's relevant to the programmer and the component.  It provides insight into where things are coming from.  I wish CanJS did this. Backbone also did this (vs `can-control`).


> 3. None of RVMs competitors, AFAIK, do this and I would like to make adoption as easy as possible (especially from redux, as it is the most popular option for state management besides "nothing", which has almost twice as many users as redux).

If "nothing" is the most popular option, we should be looking to pull people from "nothing", not from Redux. We can probably provide state management that still feels like native react.

For example, we could just make state the observable.  The following is the `Clock` example from https://reactjs.org/docs/state-and-lifecycle.html

```js
class Clock extends React.Component {
  constructor(props) {
    super(props);
    this.state = {date: new Date()};
  }

  componentDidMount() {
    this.timerID = setInterval(
      () => this.tick(),
      1000
    );
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  tick() {
    this.setState({
      date: new Date()
    });
  }

  render() {
    return (
      <div>
        <h1>Hello, world!</h1>
        <h2>It is {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}

ReactDOM.render(
  <Clock />,
  document.getElementById('root')
);
```

I think we should simply make this work:

```js
class Clock extends React.Component {
  constructor(props) {
    super(props);
    this.state = {date: new Date()};
  }

  componentDidMount() {
    this.timerID = setInterval(
      () => this.tick(),
      1000
    );
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  tick() {
    this.state.date = new Date();
  }

  render() {
    return (
      <div>
        <h1>Hello, world!</h1>
        <h2>It is {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}

ReactDOM.render(
  <Clock />,
  document.getElementById('root')
);
```

All I did was make it so instead of:

```js
this.setState({
   date: new Date()
});
```

folks can write:

```js
this.state.date = new Date();
```

=============

After CanJS 3.0, I learned hard lessons about simplicity.  While technically more correct and flexible to write out everything separate:

```js
import DefineMap from "can-define/map/map"
import stache from "can-stache";
import Component from "can-component";

var VM = DefineMap.extend({ ... })

Component.extend({
  tag: "my-tag",
  ViewModel: VM,
  view: stache(` ... `);
})
```

It wasn't as conceptually simple.  People are responding much better to the 4.0:

```js
import Component from "can-component";

Component.extend({
  tag: "my-tag",
  ViewModel: { ... },
  view: ` ... `;
})
```

It feels integrated.  

The `@connect` pattern feels like it's over-prioritizing the "supply state to an existing view" problem. Instead, I think we should solve just one problem ... having to call `.setState`.

**Comment by kylegifford** [comment]
I'd like to echo my support for @brianmoschel's proposal, and push for a marketing effort for RVM in parallel to prioritizing an architecture direction.

In working with current and future Bitovi clients, "using React" is almost guaranteed to come up in conversation. What business people (and even higher level technical managers) don't fully understand is that React is best when it's handling views/components, and that the state management framework is generally agnostic from that. They just want to "use React", and RVM gives us a chance to push the benefits of CanJS/DoneJS while still "using React".

Along with a rebranding, website, docs, explainer videos, I think it would be worth a "publicity tour": promoting this through in-person meetups in the React community, espousing the benefits of CanJS coupled with React (longevity, solid tooling within DoneJS, etc).

I'd also like @bgando, @mickmcgrath13, and @DesignByOnyx's opinions on how RVM has fared in production with CanJS to date, and any recommended changes.

**Comment by justinbmeyer** [comment]
@kylegifford IMO, we need to know the product so we can sell it the right way.  A marketing effort without knowing what the API even looks like seems wasteful.  Are we trying to peel off Redux folks, or the folks not using anything but react?  The architecture informs that.

**Comment by justinbmeyer** [comment]
Thanks for the comments everyone!

Thou ... please try to keep future discussion specific. All +1s of the concept should be saved for the upcoming survey. 

But please share specific proposals.  For example, APIs that solve stated problems, with ideas how to market the specific tool. Thanks!

**Comment by DesignByOnyx** [comment]
I'll do my best to keep this short. Most of my concerns are API related.

- Separation of props and state is a deal-breaker IMO. How it is solved doesn't matter to me, but the lack of separation has caused us problems. Reach out to me on Slack for real world examples.
- @justinbmeyer gave examples earlier which can be boiled down to "RVM provides observable state - so forget about `setState`, `forceUpdate`, and `shouldComponentUpdate`". This was my vision for `can-connect` and something I fully support.
- I personally like the class `extend` syntax. I think we can `@connect` our cake and `extend` it too without much maintenance or marketing overhead. I also think render props (re: `VMComonent`) are too advanced and cutting edge to be marketed effectively at this point.
